backend/src/main/java/com/redhat/cloud/notifications/events/KafkaMessageDeduplicator.java

```diff
@@ -0,0 +1,128 @@
+import io.micrometer.core.instrument.Counter;
+import io.micrometer.core.instrument.MeterRegistry;
+import io.smallrye.mutiny.Uni;
+import io.smallrye.reactive.messaging.kafka.KafkaMessageMetadata;
+import org.apache.kafka.common.header.Header;
+import org.eclipse.microprofile.reactive.messaging.Message;
+import org.hibernate.reactive.mutiny.Mutiny;
+import org.jboss.logging.Logger;
+
+import javax.annotation.PostConstruct;
+import javax.enterprise.context.ApplicationScoped;
+import javax.inject.Inject;
+import java.util.Iterator;
+import java.util.Optional;
+import java.util.UUID;
+
+import static java.lang.Boolean.FALSE;
+import static java.nio.charset.StandardCharsets.UTF_8;
+
+@ApplicationScoped
+public class KafkaMessageDeduplicator {
+
+    public static final String MESSAGE_ID_HEADER = "rh-message-id";
+    public static final String MESSAGE_ID_VALID_COUNTER_NAME = "kafka-message-id.valid";
+    public static final String MESSAGE_ID_INVALID_COUNTER_NAME = "kafka-message-id.invalid";
+    public static final String MESSAGE_ID_MISSING_COUNTER_NAME = "kafka-message-id.missing";
+
+    private static final Logger LOGGER = Logger.getLogger(KafkaMessageDeduplicator.class);
+
+    @Inject
+    Mutiny.StatelessSession statelessSession;
+
+    @Inject
+    MeterRegistry meterRegistry;
+
+    private Counter validMessageIdCounter;
+    private Counter invalidMessageIdCounter;
+    private Counter missingMessageIdCounter;
+
+    @PostConstruct
+    void initCounters() {
+        validMessageIdCounter = meterRegistry.counter(MESSAGE_ID_VALID_COUNTER_NAME);
+        invalidMessageIdCounter = meterRegistry.counter(MESSAGE_ID_INVALID_COUNTER_NAME);
+        missingMessageIdCounter = meterRegistry.counter(MESSAGE_ID_MISSING_COUNTER_NAME);
+    }
+
+    /**
+     * Extracts the message ID from a Kafka message header value. If multiple header values are available, the first one
+     * will be used and the other ones will be ignored. An invalid header value will be counted and logged, but won't
+     * interrupt the message processing: the deduplication will be disabled for the message.
+     */
+    public UUID findMessageId(String bundleName, String applicationName, Message<String> message) {
+        boolean found = false;
+        Optional<KafkaMessageMetadata> metadata = message.getMetadata(KafkaMessageMetadata.class);
+        if (metadata.isPresent()) {
+            Iterator<Header> headers = metadata.get().getHeaders().headers(MESSAGE_ID_HEADER).iterator();
+            if (headers.hasNext()) {
+                found = true;
+                Header header = headers.next();
+                if (header.value() == null) {
+                    invalidMessageIdCounter.increment();
+                    LOGGER.warnf("Application %s/%s sent an invalid Kafka header [%s=null]. They must change their " +
+                                    "integration and send a non-null value.", bundleName, applicationName, MESSAGE_ID_HEADER);
```

Comment:
Do we have this header for policies?

Reply:
and notifications-gw

Reply:
Ah. i see that this is something new.

Reply:
Yes it's new.

Reply:
I think almost no one is using Kafka headers in all of platform (apps)